MatrixKit/Models/Contact/MXKContactManager.m

```diff
@@ -1230,74 +1229,125 @@ + (void)requestUserConfirmationForLocalContactsSyncWithTitle:(NSString*)title
 
 #pragma mark - Internals
 
+- (NSDictionary*)matrixContactsByMatrixIDFromMXSessions:(NSArray<MXSession*>*)mxSessions
+{
+    // The existing dictionary of contacts will be replaced by this one
+    NSMutableDictionary *matrixContactByMatrixID = [[NSMutableDictionary alloc] init];
+    for (MXSession *mxSession in mxSessions)
+    {
+        // Check all existing users
+        NSArray *mxUsers = mxSession.users;
+        
+        for (MXUser *user in mxUsers)
+        {
+            // Check whether this user has already been added
+            if (!matrixContactByMatrixID[user.userId])
+            {
+                if ((self.contactManagerMXRoomSource == MXKContactManagerMXRoomSourceAll) || ((self.contactManagerMXRoomSource == MXKContactManagerMXRoomSourceDirectChats) && mxSession.directRooms[user.userId]))
+                {
+                    // Check whether a contact is already defined for this id in previous dictionary
+                    // (avoid delete and create the same ones, it could save thumbnail downloads).
+                    MXKContact* contact = matrixContactByMatrixID[user.userId];
+                    if (contact)
+                    {
+                        contact.displayName = (user.displayname.length > 0) ? user.displayname : user.userId;
+                        
+                        // Check the avatar change
+                        if ((user.avatarUrl || contact.matrixAvatarURL) && ([user.avatarUrl isEqualToString:contact.matrixAvatarURL] == NO))
+                        {
+                            [contact resetMatrixThumbnail];
+                        }
+                    }
+                    else
+                    {
+                        contact = [[MXKContact alloc] initMatrixContactWithDisplayName:((user.displayname.length > 0) ? user.displayname : user.userId) andMatrixID:user.userId];
+                    }
+                    
+                    matrixContactByMatrixID[user.userId] = contact;
+                }
+            }
+        }
+    }
+    
+    return [matrixContactByMatrixID copy];
```

Comment:
Do we really need a copy?

Reply:
It allows to convert NSMutableDictionary to NSDictionary. `copy` always returns an immutable type and `mutableCopy` returns a mutable type. Equivalent to `return [NSDictionary dictionaryWithDictionary: matrixContactByMatrixID];`